engines/grim/emi/emi.cpp

```diff
@@ -143,25 +143,48 @@ void EMIEngine::drawNormalMode() {
 	// Draw actors
 	buildActiveActorsList();
 	sortActiveActorsList();
+	sortLayers();
 
 	Bitmap *background = _currSet->getCurrSetup()->_bkgndBm;
 	background->_data->load();
 	uint32 numLayers = background->_data->_numLayers;
+
+	Common::List<Layer *>::const_iterator nextLayer = _layers.begin();
+	Common::List<Actor *>::const_iterator nextActor = _activeActors.begin();
 	int32 currentLayer = numLayers - 1;
-	foreach (Actor *a, _activeActors) {
-		int sortorder = a->getEffectiveSortOrder();
-		if (sortorder < 0)
-			break;
 
-		while (sortorder <= currentLayer * 10 && currentLayer >= 0) {
-			background->drawLayer(currentLayer--);
+	int aso = (nextActor != _activeActors.end()) ? (*nextActor)->getEffectiveSortOrder() : -1;
+	int lso = (nextLayer != _layers.end()) ? (*nextLayer)->getSortOrder() : -1;
+	int bgso = currentLayer * 10;
+
+	// interleave actors, background layers and additional stand-alone layers based
+	// on their sortorder
+	//
+	// priority for same sort order:
+	//   background layers (highest priority)
+	//   stand-alone layers
+	//   actors
+	while (1) {
```

Comment:
I feel like this block could be refactored to be much nicer, but I can't think of how. It just feels weird not having a loop condition.